common/src/main/java/io/netty/util/AsciiString.java

```diff
@@ -533,14 +533,19 @@ public boolean contentEqualsIgnoreCase(CharSequence string) {
             return false;
         }
 
+        byte[] value = this.value;
         if (string instanceof AsciiString) {
             AsciiString rhs = (AsciiString) string;
-            for (int i = arrayOffset(), j = rhs.arrayOffset(), end = i + length(); i < end; ++i, ++j) {
-                if (!equalsIgnoreCase(value[i], rhs.value[j])) {
-                    return false;
+            byte[] rhsValue = rhs.value;
+            if ((arrayOffset() | rhs.arrayOffset()) == 0) {
+                for (int i = 0; i < length; ++i) {
+                    if (!equalsIgnoreCase(value[i], rhsValue[i])) {
+                        return false;
```

Comment:
Will it benefit JIT more if we extract this path into its own method (like `alignedEqualsIgnoreCase`) and make both methods static?

Reply:
I preferred to have the common case inlined and kept the size (in bytecode), of the callee small enough to make it inlineable as well. It's a type of pattern we use in different part of Netty, which is coming from OpenJDK "guidelines" for inlining afaik

Reply:
@franz1981 sounds good... do you have a link to the OpenJDK guidelines ? would be interested to read up here.

Reply:
yep are old but still decent ->  https://wiki.openjdk.org/display/HotSpot/Inlining

> There are several ways to help the inliner make the decision you want it to:
> - ...
> - Segregate fast paths from slow paths, keeping hot methods small (35 bytecodes or less) with error processing and slow paths in separate methods.

This one is the one I usually follow; the `equasIgnoreCase` is small enough (verified by inspecting the assembly) while the "common" expected path is the one I've decided to inline, keeping the other "slower" ones separated.